### ISSUE TITLE
Discuss SSR API and methods

### ISSUE THREAD

**Issue by thisguychris** [styled-components/styled-components#386]
I just want to kick off a discussion to finalise our SSR API.

So API came from the comments on the #214 

```JS
const { css, html } = StyleSheet.collect(() =>
  ReactDOMServer.renderToString(<App />)
);
```

Please feel free to suggest what the API will look like, (let the bikeshedding begin!)

Also if we're already doing de-duping in SSR level, do we even need reset()? (related issue #378)  Or would that just be a convenience function since we don't need it anymore?

cc @mxstbr / @geelen

**Comment by acjay** [comment]
One thing I would throw out there is that it would be very useful to be able to have separate bundles of CSS. That way, you could target styles to a `style` tag or to a separate stylesheet. When I worked for a publisher, we used this type of optimization to achieve very fast initial renders, with nonessential styles deferred until later.

I can ticket this separately, if that would be helpful. It would require some sort of ability to annotate at the component or CSS level where rulesets would be targeted, and an API for collecting the separate bundles (which might be a `collect` variant).

**Comment by mxstbr** [comment]
This happens either way @acjay. With styled-components, only the styles of components that are actually used on the initial render are injected 👍 

Generating `.css` files for server-side rendered apps would not be complicated to add, but should not live in core. (it's more or less just a `fs.write(css)` away) We're going to work on a babel or webpack plugin to make this possible for non-ssr apps but that's a different issue. (see #59)

**Comment by dfrankland** [comment]
@thisguychris, `reset()` is great for doing testing. Considering that tagged templates have logic in them, that logic may need to be tested.

**Comment by mxstbr** [comment]
That's a great point actually @dfrankland, I hadn't even considered that use case!

**Comment by nicksrandall** [comment]
We will also need to keep track of the classNames that are generated so we can rehydrate them on the client. 

For example, this is how glamor does it... https://github.com/threepointone/glamor/blob/master/docs/server.md

**Comment by camsjams** [comment]
Looks good to me

**Comment by nicksrandall** [comment]
Do we also want to be able to export class names here ( in addition to CSS and HTML) like Aphrodite and Glamor does so that we can consume them on the front end to make sure things stay in sync or are just going to rely on the babel plugin to ensure deterministic class names?

**Comment by xcoderzach** [comment]
Even if the names are deterministic, you still want a list of the class names, so you don't inject every rule twice.  Either that, or maybe we could have a `class` attribute on `style` and `link`s like `__styled-component-sheet__`, and then just build the list from those sheets.

**Comment by wmertens** [comment]
I don't think we need to worry about duplicate classnames after SSR; just remove the server-provided style block after the app code arrived and let S-C rebuild the styles.

**Comment by kuka** [comment]
I like what Aphrodite does. Just pass the classnames along with the initial state and rehydrate without rebuilding client-side. Does anyone see any cons with this approach?

**Comment by nicksrandall** [comment]
@wmertens we do need to worry about duplicate class names because react will check the root elements html on its first render and see if it is the same. If it is then react will resuse it, otherwise it will blow it away and start from scratch (thus defeating the purpose of srr). If the class names are differeent the that first check won't pass.

**Comment by wmertens** [comment]
Oh right that's a different issue, that should be handled by the Babel
plugin.

Actually, I am running v2 with the plugin and it doesn't work for me, the
generated classnames change between server and client.

I don't understand how the names can differ though; given that the entire
render result is normally the same, and the order of evaluation thus also,
shouldn't the generated names be the same as well? If you use a
deterministic naming scheme, that is?

**Comment by ForbesLindesay** [comment]
@wmertens I found that in order to get the names to match, I had to make sure each file was transpiled via babel once, and then the result re-used on client and server.  I was initially applying babel twice (once for the server side and once for the client side).  This was resulting in different names.

It would be cool if in the babel plugin, we could make the names deterministic (e.g. use an md5 hash of the file's source)

**Comment by wmertens** [comment]
:( I can't do that, the Babel options for client and server differ…
Strangely, before using the plugin, the names given by S-C were fine…

**Comment by GGAlanSmithee** [comment]
I am experiencing the same @wmertens

**Comment by ForbesLindesay** [comment]
I think possibly the simplest option might be some kind of `.styled-components-cache.json` file that could map sha hashes of files + index of styled component to generated class name.  That way we could ensure they were always deterministic, without having to make them long enough to avoid collisions.

On the other hand, if they're currently just random IDs, maybe just converting them to the first few letters of the sha has would be good enough.

**Comment by wmertens** [comment]
For SSR, all you need is "the first one, the second one, ..." since on the
client they will be encountered in the same order. So class names can just
be a prefix with a number.

Right?

**Comment by xcoderzach** [comment]
@wmertens There are a lot of ways the client and server execution could be different.  And I think it's unreasonable to force that restriction on people.

e.g.

```
if(isBrowser) {
  styled.div``
}
```

could potentially break everything.

**Comment by ForbesLindesay** [comment]
Yes, but as @xcoderzach points out, you can declare components that will be used on the second render of the client side code, but that you don't need on the server side.

The babel plugin is definitely the right approach.  I think the only possible issue is that I don't think we can rely on files always being processed in the same order by babel.  e.g. `babel src --out-dir lib` would be very different to `browserify src/index.js -t babelify`.  This is why I don't think our solution should rely on random number generation or the order in which files are processed.

**Comment by vdanchenkov** [comment]
@ForbesLindesay I'm on making id's deterministic.

**Comment by oliviertassinari** [comment]
I have been asked by @mxstbr to comment on that issue.
The *block* API proposed by @satya164 looks definitely simpler than the `getCSS()`/`reset()` one. The teardown is handled for you. 

However, my concern with those two APIs is around request isolation and concurrency. Relying on a singleton to collect the style prevents from handling two requests concurrently.

We have been implementing a different API with JSS on the new version of [Material-UI](https://github.com/callemall/material-ui/tree/next). We instantiate a style collector at each request.
This example is taken from the Server Rendering [documentation](https://material-ui-1dab0.firebaseapp.com/#/getting-started/server-rendering).
```jsx
function handleRender(req, res) {
  // Create a styleManager instance.
  const { styleManager, theme } = createStyleManager();

  // Render the component to a string.
  const html = renderToString(
    <MuiThemeProvider styleManager={styleManager} theme={theme}>
      <App />
    </MuiThemeProvider>
  )

  // Grab the CSS from our styleManager.
  const css = styleManager.sheetsToString()

  // Send the rendered page back to the client.
  res.send(renderFullPage(html, css))
}
```

To be fair, that API only shine when using a [streaming rendering](https://github.com/aickin/react-dom-stream) implementation that can be interrupted. When using the synchronous `ReactDOMServer.renderToString()` implementation, I don't think that it makes much different. Maybe, it's simpler with unit tests.

**Comment by wmertens** [comment]
On Mon, Feb 13, 2017 at 1:32 PM Olivier Tassinari <notifications@github.com>
wrote:

> To be fair, that API only shine when using a streaming rendering
> <https://github.com/aickin/react-dom-stream> implementation that can be
> interrupted. When using the synchronous ReactDOMServer.renderToString()
> implementation, I don't think that it makes much different. Maybe, it's
> simpler with unit tests.
>

I like it! Besides, people can decide to use streaming rendering with S-C,
so definitely a good thing.

**Comment by oliviertassinari** [comment]
Looks like asynchronous SSR is now a real thing 🎉  https://github.com/FormidableLabs/rapscallion.

**Comment by alloy** [comment]
I would also very much like to be able to use this in an asynchronous streaming manner.

**Comment by siddharthkp** [comment]
Landed here while debugging a rapscallion + styled-components implementation.

Does anyone have a working example 🙈

@mxstbr?

**Comment by wmertens** [comment]
I think https://github.com/FormidableLabs/rapscallion/issues/39 is probably
a better issue for that question?

**Comment by wjagodfrey** [comment]
`nfl/react-helmet` (head element side-effect component) have a single method call to handle server side rendering. Using it in a large production application with great success.
https://github.com/nfl/react-helmet#server-usage

**Comment by mxstbr** [comment]
With React v16 (Fiber) vanilla streaming SSR support is on the roadmap, so we really want an API that works with that.

The classes from the Babel plugin are now deterministic. The TL;DR of this discussion, as far as I can tell, is that we need a **non-singleton for concurrency** and a **single method** rather than two separate ones.

This doesn't have to be too different from what we already have. If we export a non-singleton that means we can still have the `getCSS` and `reset` methods, but you won't have to use `reset` on the server since it's not a singleton anyway. (still useful for testing though)

Taking that into account the API would look something like this:

```JS
import { getStyleSheet } from 'styled-components';

function route(req, res) {
  // Get the styleSheet for this request
  const styleSheet = getStyleSheet();
  
  // Render the app
  const html = renderToString(<App />);

  // Get the critical CSS
  const css = styleSheet.getCSS();

  // Send the response
  res.send(render(html, css));
}
```

Does this cover everything? If I'm thinking about this correctly, this is just a tiny change to add a function wrapper around our `styleSheet` export.

**Comment by alloy** [comment]
@mxstbr With that example, how would the `styleSheet` instance be used during `renderToString` if it’s not passed in anywhere? I think most solutions put the instance in the React Context, so it’s available everywhere, but I don’t see that being done in your example.

**Comment by alloy** [comment]
Wait… am I crazy? Where did that example that I received an email about go? 🤔

**Comment by mxstbr** [comment]
With React v16 (Fiber) vanilla streaming SSR support is on the roadmap, so we really want an API that works with that.

The classes from the Babel plugin are now deterministic. The TL;DR of this discussion, as far as I can tell, is that we need a **non-singleton for concurrency** and a **single method** rather than two separate ones.

This doesn't have to be too different from what we already have. If we export a non-singleton that means we can still have the `getCSS` and `reset` methods, but you won't have to use `reset` on the server since it's not a singleton anyway. (still useful for testing though)

The one thing that needs to change is that we need to keep track of which CSS was injected in that request. I quite like the API that was suggested by @oliviertassinari (thanks for chiming in btw), though I don't think the `ThemeProvider` should care about the style manager, that should be its own thing imo.

Taking all of those thoughts into account the API would look something like this:

```JS
import { getStyleManager } from 'styled-components';

function route(req, res) {
  // Get the styleSheet for this request
  const { StyleManager, styleSheet } = getStyleManager();
  
  // Render the app
  const html = renderToString(
    <StyleManager styleSheet={styleSheet}>
      <App />
    </StyleManager>
  );

  // Get the critical CSS
  const css = styleSheet.getCSS();

  // Send the response
  res.send(render(html, css));
}
```

Does this cover everything? The big change to the current API would be that the styleSheet can either be constructed client-side or come from context.

**Comment by alloy** [comment]
Aha, I guess you deleted it and added a new one that _does_ show the instance being injected :) 👍 

One question re streaming, though, it looks like with this example the full render has to be performed before sending any data to the client, how would the streaming version work?

**Comment by mxstbr** [comment]
> One question re streaming, though, it looks like with this example the full render has to be performed before sending any data to the client, how would the streaming version work?

That depends on how React Fiber implements it, I have no idea! Probably something like this?

```JS
  // Render to stream
  const htmlStream = renderToStream(
    <StyleManager styleSheet={styleSheet}>
      <App />
    </StyleManager>
  );

  // Get the critical CSS
  const css = styleSheet.getCSS();

  // I have no idea how this would work?
  res.send(injectCssIntoStream(htmlStream, css));
```

How would one inject the CSS into a stream of HTML? Maybe we do need a different API?

**Comment by alloy** [comment]
To be clear, I have no idea either and no previous experience with this. Conceptually it seems tough as we'd want the styling in the head element, but the CSS isn't known until rendering is done. Maybe this can only be done with ahead of time compilation of the CSS or caching it after the first request or something?

Are there examples of other frameworks that already do this?

**Comment by mxstbr** [comment]
>  Conceptually it seems tough as we'd want the styling in the head element, but the CSS isn't known until rendering is done. 

You are right, that is going to be a pain. At least the new API makes concurrency possible, I don't know if streaming SSR is ever going to work with css-in-js though, given we don't know which CSS to inject until everything is rendered... :confused: /cc @trueadm @gaearon

**Comment by alloy** [comment]
> At least the new API makes concurrency possible

True 👍 

> I don't know if streaming SSR is ever going to work with css-in-js though

Inline styles would probably work, but frameworks that collect styles and want to inject into head seem to be impossible yeah.

Is the babel plugin still meant to be able to extract all styling and serve that as a precompiled stylesheet? I guess that might be the only way to add streaming.

**Comment by mxstbr** [comment]
Talking to @trueadm it's pretty clear for streaming rendering support we'll need to have a static CSS file. The bable/webpack plugin is going to get us pretty far, but not all the way there.

Rather, we'll need to simulate as if a real style sheet existed with the generated styles from the render. This would be done by dynamically injecting a `<link>` into the streamed HTML which points towards the right CSS in the cache.

Basically, it'd look something like this: (note, pseudo code)

```JS
const styleCache = {};

app.get('/stylesheet.css?:id', (req, res) => {
  // Pretend as if a style sheet exists here and send back the generated styles for this request
  res.send(styleCache[req.params.id]);
})

app.get('*', (req, res) => {
  const { StyleManager, styleSheet } = getStyleManager();
  const id = generateUniqueHash();
  renderToStream(
    // This renders a <link href=`/stylesheet.css/?${id}` />
    <StyleManager styleSheet={styleSheet} styleId={id}>
      <App />
    </StyleManager>
  ).pipe(res);
  // Cache the CSS of this request after the stream has finished rendering
  // TODO: Figure out how to do this after the stream has finished rendering, for now let's
  // assume it's synchronous
  styleCache[id] = styleSheet.getCSS();
})
```

**Comment by mxstbr** [comment]
As proposed above, this would be the full API extension for `styled-components`:

```JS
import { getStyleManager } from 'styled-components';

const { StyleManager, styleSheet } = getStyleManager();

<StyleManager styleSheet={styleSheet} />

styleSheet.getCSS();
styleSheet.reset();
```

**Does anybody have input regarding naming or use cases this doesn't cover?** (personally not a fan of the word `StyleManager`, maybe we can think of something better to describe it?) Otherwise I say we go ahead and start implementing a WIP of this!

**Comment by satya164** [comment]
Sorry, I'm a bit out of loop here

> @oliviertassinari : However, my concern with those two APIs is around request isolation and concurrency. Relying on a singleton to collect the style prevents from handling two requests concurrently.

What I proposed doesn't really on a singleton, it can just manage creating new instance internally instead of exposing it.

What do you mean by handling two requests concurrently? Can you give an example? 

> @mxstbr 

Wondering what's the advantage of the `reset` method as opposed to just creating a new styleSheet object if you need, also why do we need a new instance of `StyleManager` each time? Won't `<StyleManager />` give us a new react element every time?

```js
import { StyleManager, StyleSheet } from 'styled-components';

const styleSheet = StyleSheet.create();

<StyleManager styleSheet={styleSheet}><MyApp></StyleManager>

styleSheet.getCSS();
```

Though I'm wondering, what's the purpose of the `styleSheet` element. Where else it will be used? Am I going to use it to render multiple elements and collect css from each of them?

**Comment by mxstbr** [comment]
> Wondering what's the advantage of the reset method as opposed to just creating a new styleSheet object if you need

Makes testing easier:

```JS
import { styleSheet } from 'styled-components'

beforeEach(() => {
  styleSheet.reset();
})
```

> why do we need a new instance of StyleManager each time? Won't <StyleManager /> give us a new react element every time?

That's a good point, I quite like your API!

The purpose of the separate `styleSheet` object is to avoid having two request referring to the same CSS when concurrently rendering, what @oliviertassinari said. (I have no idea is anybody is doing that or if that problem exists, I don't really do SSR, but it seems like a likely issue to come up)

**Comment by satya164** [comment]
@mxstbr

Aren't these almost same? I don't like extra methods xD

```
const styleSheet = createStyleSheet() // `StyleSheet.create()` or `getStyleManager()`

beforeEach(() => {
  styleSheet.reset();
})
```

vs

```
let styleSheet = createStyleSheet() // `StyleSheet.create()` or `getStyleManager()`

beforeEach(() => {
  styleSheet = createStyleSheet()
})
```

> The purpose of the separate styleSheet object is to avoid having two request referring to the same CSS when concurrently rendering

Yeah, I get that. I was asking what's the purpose of making the user manually create the instance as opposed to automatically handling it internally. e.g. -

```
const { css, html } = StyleSheet.collect(styleSheet =>
  ReactDOMServer.renderToString(<App styleSheet={styleSheet} />)
);
```

**Comment by mxstbr** [comment]
> Aren't these almost same?

Yeah you're right, if we go with `StyleSheet.create()` that does the same thing.

How would your API work with streaming SSR? (i.e. `styleCache`)

**Comment by satya164** [comment]
@mxstbr `html` and `css` could be streams instead of strings perhaps, the css gets populated as new html is rendered. Though I don't have experience with streaming rendering, so I'm not totally sure.

**Comment by mxstbr** [comment]
You would probably do something like this:

```JS
const { StyleSheet, StyleManager } = getStyleManager();

const { css } = StyleSheet.collect(styleSheet => {
  renderToStream(<StyleManager styleSheet={styleSheet}><App /></StyleManager>).pipe(res);
})
```

Hmm, I quite like that actually as it forces people to work concurrently...

**Comment by geelen** [comment]
Ok just catching up here. Let me know if I've missed anything, but it seems like this would work:

```js
import { StyleSheet } from 'styled-components'

const { css, html } = StyleSheet.collect(withStyleSheet => (
   ReactDOMServer.renderToString(withStyleSheet(<App/>))
))
```

This potentially lends itself to some kind of streaming API:

```js
import { StyleSheet } from 'styled-components'

const css = new Stream()
const html = new Stream()

StyleSheet.toStream(withStyleSheet => {
  renderToStream(withStyleSheet(<App/>)).pipe(html)
}).pipe(css)
```

Personally, I prefer the idea of a HOC for attaching a stream rather than a `<StyleManager>`.

Thoughts?

**Comment by wmertens** [comment]
It's not HoC because it doesn't transform a component, but does something
with an element. However, I guess it works as long as the resultant React
tree is the same as on the client.

**Comment by philpl** [comment]
Frankly, I don't really see the point of streaming the css bit. It's being built up gradually as the components are rendered, and we want it to be injected at the bottom of the HTML then, I guess? So by the time we reach that bit of the HTML, the css should be complete and doesn't need to be streamed, right?

**Comment by alloy** [comment]
@mxstbr I don’t see how the [suggested cached CSS approach](https://github.com/styled-components/styled-components/issues/386#issuecomment-290917359) would really work. The CSS would still only be complete by the end of rendering the full tree, so when _streaming_ the HTML to the browser it would fail to fetch the CSS once it’s received the head element, no?

> The babel/webpack plugin is going to get us pretty far, but not all the way there.

What would it not be able to cover?

**Comment by mxstbr** [comment]
> However, I guess it works as long as the resultant React tree is the same as on the client.

The DOM will be the same, there'll only be an _element_ more on the server. (`<StyleManager />`) That shouldn't be an issue, right?

I like that API @geelen, less surface area is great!

Not a fan of `StyleSheet.collect(withStyleSheet)` in terms of naming though, as first we're collecting a stylesheet but then adding another stylesheet? Sounds pretty confusing. What about this instead:

```JS
import { StyleSheet } from 'styled-components'

const { css, html } = StyleSheet.extract(collectStyles => (
   ReactDOMServer.renderToString(collectStyles(<App/>))
))
```

**Comment by mxstbr** [comment]
> so when streaming the HTML to the browser it would fail to fetch the CSS once it’s received the head element, no?

Yeah, you're right, you'd need to inject the `<link />` at the end of the doc.

> What would it not be able to cover?

Anything dynamic will have to stay at runtime, so we can't just use the webpack plugin and be done with it.

**Comment by wmertens** [comment]
The React tree has to be the same on client and server for rehydration.

Also, if the CSS is at the bottom of the page, there will be a FoUC unless
you have a rule that hides the page until after the CSS loaded? For a long
page (e.g. blog post with comments), that could be a long wait.
Alternative solution: embed <style> tags wherever they are created, Just In
Time CSS. I believe this is what jsx-style does.

After the render hydrated, these tags could be dropped from the tree with
the CSS added to head, that should not cause re-layout.

**Comment by mxstbr** [comment]
> The React tree has to be the same on client and server for rehydration.

Are you sure? If that's the case the whole idea won't work.

> embed <style> tags wherever they are created, Just In Time CSS. I believe this is what jsx-style does.

Interesting, how would you do that? This could be really slow, right?

**Comment by wmertens** [comment]
Ah it was https://github.com/zeit/styled-jsx#how-it-works

and yes, I'm pretty sure. The React tree has to be the same from SSR and
first client render, because otherwise the react id's don't match. So to
make withStylesheet work, you'd either need to do some internal-React-API
context hackery on the provided <App/>, or simply return it wrapped in a
<StyleManager> :)

On the client, you'd then also need a <StyleManager/>, at which point it is
just the same as a Provider.

**Comment by philpl** [comment]
@mxstbr @wmertens another problem, courtesy of @nfcampos

Are we taking into account that 2+ renders can take place at once? Do we want to introduce a `StyleSheetProvider` to solve this, or something?

**Comment by nfcampos** [comment]
@philpl to clarify, 2+ renders can happen at the same time if using an asynchronous SSR solution, e.g. https://github.com/FormidableLabs/rapscallion)

**Comment by alloy** [comment]
@philpl @nfcampos That’s what the concurrency and getting rid of singletons parts in this thread refer to.

**Comment by philpl** [comment]
So to summarise, we need to:

- write a `<StyleManager styleSheet={styleSheet}>` provider
- find out how to stream the css at the same time-ish?

The streaming in parallel seems problematic to me, while the second provider should be simple

cc @nfcampos

**Comment by wmertens** [comment]
How about giving ThemeProvider an optional `styleSheet` to inject? That
way, you can use existing API, and people can attach their theme at the
same time (optionally of course)

On the client, you don't inject the stylesheet, or you get the singleton
from the module, either way works. So then it would be

import { StyleSheet } from 'styled-components'
const { css, html } = StyleSheet.extract(sheet => (
   ReactDOMServer.renderToString(<ThemeProvider
sheet={sheet}><App/></ThemeProvider>))
))

**Comment by mxstbr** [comment]
> Are we taking into account that 2+ renders can take place at once?

That's exactly what this whole discussion is about, and why our current API idea doesn't work :wink:

> find out how to stream the css at the same time-ish?

I don't think this is necessary for now, we can talk about this once we know what the React API looks like.

> How about giving ThemeProvider an optional `styleSheet` to inject? That way, you can use existing API, and people can attach their theme at the same time (optionally of course)

That doesn't solve the "react tree is different on client and server"-issue though, does it?

**Comment by wmertens** [comment]
> That doesn't solve the "react tree is different on client and
server"-issue though, does it?

Yes it does, same components, same tree, only a different styleSheet is
injected in context.

**Comment by mxstbr** [comment]
> Yes it does, same components, same tree, only a different styleSheet is injected in context.

I don't understand.

This is our client `App.js` that's the root element rendered on the page:

```JS
const App = () => {
  return (
    <ThemeProvider theme={x}>
      <Root />
    </ThemeProvider>
  )
}

ReactDOM.render(<App />, node);
```

On the server we then do this to inject the `styleSheet`:

```JS
renderToString(<ThemeProvider styleSheet={styleSheet}><App /></ThemeProvider>)
```

Wouldn't that result in two different trees?

```JS
// client:
<ThemeProvider theme={x}>
  <Root />
</ThemeProvider>

// server:
<ThemeProvider styleSheet={styleSheet}>
  <ThemeProvider theme={x}>
    <Root />
  </ThemeProvider>
</ThemeProvider>
```

**Comment by geelen** [comment]
Yeah ok I think let's ditch any discussion of streaming for now and focus on a single use-case, then we can discuss changes to better support Rapscallion/Fiber in another issue, when we have a clearer idea of their constraints.

**Comment by nfcampos** [comment]
i think i agree with glen but i think we should still consider supporting asynchronous (not streaming) rendering now (i.e. not use a singleton)

**Comment by geelen** [comment]
@nfcampos yep exactly. Avoiding singletons is a must.

**Comment by wmertens** [comment]
Ah, sorry for being unclear. I meant that the ThemeProvider would be in both client and server, and probably it would look something like this:

This is our client App.js that's the root element rendered on the page:
```js
const App = ({styleSheet}) => {
  return (
    <ThemeProvider styleSheet={styleSheet} theme={x}>
      <Root />
    </ThemeProvider>
  )
}

ReactDOM.render(<App />, node);
```

On the server we then do this to inject the styleSheet:
```js
renderToString<App styleSheet={styleSheet} />)
```

so the trees are the same on client and server, only on the client we don't provide a `styleSheet` which makes it default to the module's singleton. (or we do provide one, doesn't matter)

**Comment by geelen** [comment]
As for the server/client differences, the way our ThemeProvider works is that it [renders an only child](https://github.com/styled-components/styled-components/blob/master/src/models/ThemeProvider.js#L80):

```js
  render() {
    if (!this.props.children) {
      return null
    }
    return React.Children.only(this.props.children)
  }
```

What does this look like after `renderToString`? Obviously, they're not rendering any markup, but does their presence change the checksum?

I'm actually slightly leaning away from my earlier syntax suggestion, though I'd be happy to leave it in as a shorthand:

```js
import { StyleSheet } from 'styled-components'

// Shorthand for straightforward case
const { css, html } = StyleSheet.extract(collectStyles => (
   renderToString(collectStyles(<App/>))
))

// Longer form in case your rendering is more complex than StyleSheet.extract can handle
const sheet = StyleSheet.create()
const html = renderToString(sheet.attach(<App/>))
const css = sheet.css
```

But I don't like `<App styleSheet={styleSheet}/>` (where you have to manage it yourself) or `<ThemeProvider styleSheet={styleSheet}/>` (an extra layer).

**Comment by geelen** [comment]
I suppose you could actually do something like this:

```js
const Sheet = StyleSheet.createInMemoryStyleSheet()
const html = renderToString(<Sheet><App/><Sheet/>)
const css = Sheet.css
```

Starting to feel a bit React Router-y, but you could then have:

```js
const Sheet = StyleSheet.createStyleTagInHead()

ReactDOM.render(<Sheet><App/><Sheet/>)
```

on the client (optionally, of course). The good thing about _this_ would be that we could expose stylesheet lifecycle hooks, so you could have some control over which style tags get used. Which has been asked for in #235 and would help #568, and maybe even #622?

```js
const Sheet = StyleSheet.createInlineStyleTag()

const Comp = styled.div`
  ...
`

export default props => (
  <div>
    <textarea onChange={ () => Sheet.reset() }/>
    <Sheet>
      <Comp/>
    </Sheet>
  </div>
)
```

Again, my only worry is making this _completely_ optional, so the current zero-config approach still works until you want some advanced use-case.

**Comment by wmertens** [comment]
Argh of course you are right, only the actually emitted tree is what
counts. I eat my words, sorry 🙊.

That said, I like this <Sheet> API you propose, and you can make it
optional by having a default singleton Sheet that gets overridden by the
wrapping Sheet.

Not so sure about the inMemoryStyleSheet, is there any other kind? :)

How about

import {Sheet} from 'styled-components'

const app = <Sheet><App/><Sheet/>

const html = renderToString(app)
const {css} = app

?

So, the css is stored on the created Sheet element. If the Sheet is
instantiated inside a browser, it will put the style on `<head>`,
otherwise it will keep it on `this.css`.

**Comment by wmertens** [comment]
Oh, InMemoryStyleSheet vs InBrowserStyleSheet, of course. Is it necessary
to make this distinction if the Sheet automatically does the right thing
depending on environment?

And you propose putting the css on a React component instead of on the
element, making a new component every time. That makes it easier to get the
generated CSS since you don't have to keep a reference to the created Sheet
element, but it feels a little non-React-y to me.

**Comment by mxstbr** [comment]
That's what I thought, so this whole discussion went in the totally wrong direction :wink:

So this API actually works out perfectly fine:

```JS
const { css, html } = StyleSheet.extract(collectStyles => (
  renderToString(collectStyles(<App />))
))
```

Let's go with that? It's definitely the easiest for the user, as the client-side code doesn't have to change at all.

With the `StyleSheet` API everything gets a bit muddy and complex, I don't think we really need that?

**Comment by kuka** [comment]
Overall, not sure if streaming even worth looking into it at this point.

@mxstbr looks good for me.

**Comment by geelen** [comment]
Ok so I think definitely a `StyledComponent` needs to look up the instance of `StyleSheet` from context (defaulting to the current singleton if it's not present). Then we can do SSR, concurrent rendering, etc, but it _also_ makes my `<Sheet>` object possible too.

I'm not sold on the `StyleSheet.extract(collectStyles =>` API though, doesn't read all that well. Imo, if we're going for a shorthand, it should be real short:

```js
import { renderToHtmlAndCSS } from 'styled-components'

const { css, html } = renderToHtmlAndCSS(<App />)
```

i.e. hide `ReactDOMServer.renderToString`. If, for some reason, that shorthand doesn't work for people, I think a more imperative API makes more sense:

```js
import { ServerSheet } from 'styled-components'

const sheet = new ServerSheet()
const html = ReactDOMServer.renderToString(sheet.extractStyles(<App/>))
const css = sheet.css
```

Alternatively, do:

```js
import SSR from 'styled-components/ssr' // or even its own package?

/* shorthand */
const { css, html } = SSR.render(<App />)
/* imperative */
const sheet = SSR.createSheet()
```

**Comment by mxstbr** [comment]
I dislike hiding the `renderToString`, if more than a single library you use do that you're f***ed, and I don't want to ship two different APIs to do the same thing.

I like you `<Sheet>` component, so why don't we do this?

```JS
import { StyleSheet } from 'styled-components';

const { html, css } = StyleSheet.extract(Sheet => renderToString(<Sheet><App /></Sheet>));
```

I don't feel strongly between this and the `collectStyles` HoC that does the same thing automatically.

**Comment by geelen** [comment]
Argh crap I've hit a wall with #679. There's a fundamental disconnect between Styled Components and the needs of SSR. Basically, the problem is this:

```js
import React from 'react'
import { renderToString } from  'react-dom/server'
import styled, { ServerStyleSheet, injectGlobal } from 'styled-components'

const sheet = new ServerStyleSheet()

const Heading = styled.h1`
  color: red;
`
injectGlobal`
  body { background: papayawhip; }
`

const html = renderToString(sheet.collectStyles(<Heading>Hello SSR!</Heading>))
const css = sheet.css
```

The way SC works at the moment, `styled.h1` and `injectGlobal` ***both generate CSS synchronously***, so only the `color: red;` rule actually gets picked up in the output. In the case of `styled.h1`, it's `.sc-component-id {}`, which has the side-effect of *preserving declaration order in the resulting CSS*. I've talked about why order is of paramount importance before so I don't want to repeat myself here, so let's just focus on `injectGlobal`.

**How can we make `injectGlobal` aware of the `sheet` that we're building up?**

There's no React component hierarchy, so no context we could use. `injectGlobal` itself is a singleton, so we only have singleton options:

```js
const sheet = new ServerStyleSheet()
styled.StyleSheet.instance = sheet // this would be hidden from the user in practice, probably

injectGlobal`
  body { background: papayawhip; }
`
```

So, we can't support concurrent SSR without making breaking changes to this API. Perhaps `injectGlobal` gets replaced with something that returns a component, so it has to be in the React render tree somehow?

```js
const Heading = styled.h1`
  color: red;
`
const BaseStyles = globalCSS`
  body { background: papayawhip; }
`

const App = () => (
  <div>
    <BaseStyles/>
    <Heading>Hello SSR!</Heading>
  </div>
)
```

Anyway, this is a blocker, so... any suggestions welcome!

**Comment by satya164** [comment]
Instead of `injectGlobal`, how about a `<GlobalStyles />` component?

**Comment by alloy** [comment]
Yeah something like that will be needed and seems fine to me. API-wise I’d probably like a HOC approach for this, but not sure if that’s what you’re mulling over or just how to approach the non-singleton implementation at all?

```js
const App = () => (
  <div>
    <Heading>Hello SSR!</Heading>
  </div>
)

export globalCSS(App)`
  body { background: papayawhip; }
`
```

**Comment by acjay** [comment]
I've followed this from a distance, so I'm not sure how all the pieces fit together. It sounds like the object `new ServerStyleSheet()` creates holds all the state of this process. Could `injectGlobal` be a method on that object?

**Comment by alloy** [comment]
@acjay The stylesheet instance would be injected into React’s ‘context’ property, so that isn’t available until a tree is being built.

**Comment by acjay** [comment]
I see how `context` is being used to provide access to the accumulating stylesheet from the component tree, but it seems to me (naively perhaps) that the underlying state could exist outside of `context` and also be accessible via methods on the stylesheet instance itself, outside of its React component interface .

**Comment by alloy** [comment]
I don’t disagree with that, the React component interface isn’t needed at all for global CSS, it’s just the most obvious way to pass the stylesheet instance along to all the modules (by piggybacking).

**Comment by mxstbr** [comment]
I just thought of a different API for `injectGlobal` yesterday during showering that could solve this issue:

```JS
import { GlobalStyle } from 'styled-components';

// ...

<GlobalStyle>
  body,
  html,
  #root {
    width: 100%;
  }
</GlobalStyle>
```

By putting our global styles back into JSX we can use context again?

**Comment by satya164** [comment]
That sounds perfect @mxstbr

**Comment by wmertens** [comment]
OTOH, maybe globalstyles can be global, as in, they get injected everywhere?

**Comment by satya164** [comment]
That can be confusing sometimes. IMO a component feels more natural and is explicit where you want it.

**Comment by geelen** [comment]
The problem with that @mxstbr (and my suggestion too) is that global styles _aren't_ components. They don't render any HTML markup, they only cause a side-effect on the whole document. There's no sense in which they could be _umounted_ (except by deleting the CSS they injected). If you wanted to do global styles in a component-centric way, you could do:

```js
const Root = styled.html`
  /* effectively global styles here */
`

React.render(<Root/>, document.documentElement)
```

But it's ok, I think I've figured out my problem. If you consider SSR breaking in to two passes, setup (where global injection happens, components are constructed in order) and render (branching down the React tree and producing CSS), then it's fine for the first step to be a singleton. Reproducing my example from before:

```js
import React from 'react'
import { renderToString } from  'react-dom/server'
import styled, { ServerStyleSheet, injectGlobal } from 'styled-components'

/* Setup */
const Heading = styled.h1`
  color: red;
`
injectGlobal`
  body { background: papayawhip; }
`

/* Render */
const sheet = new ServerStyleSheet()
const html = renderToString(sheet.collectStyles(<Heading>Hello SSR!</Heading>))
const css = sheet.css
```

When you call `new ServerStyleSheet()`, you clone the current singleton and pass it down to the components, so each render stage is separate. But, importantly, when the app boots, by virtue of there being no `global.document`, the default `StyleSheet.instance` is an in-memory one that simply buffers all its rules into its `.css` property.

So, no breaking changes needed, just a bit more messing around.

**Comment by wmertens** [comment]
What about lazy-loaded components that inject styles on second render for
data-loading SSR? Wouldn't those go into parallel singletons?

I think it's safer to require globals css to be a component without
wrapping class and treat it like all the rest, like proposed above.

**Comment by geelen** [comment]
No, that's fine. The point is that `injectGlobal` will always use the singleton under the hood. So, if you have lazy-loaded components injecting global styles, they'll always just keep adding to the singleton's buffer of global styles. The same way they would lazily apply if you used bundle-splitting on the client and had global styles spread around. On the server, any render calls _after_ that file is lazy-loaded will have these new styles included.

Obviously, side-effects in lazy-loaded files, particularly ones that have global effects (rather than just adding more animation keyframes, for example), is a bad idea. But we don't need to make the API more cumbersome, and less consistent, by trying to shoehorn global styles into a component, when it really isn't.

My `styled.html` example above already works, and it would work just as well with a `styled.div` at the root of your project. If you want to attach your "global" styles using a component, then use one of those. Imo, truly global styles are something different.

**Comment by wmertens** [comment]
Ok fair enough - I only use global styles to reset css and stretch the body
to the screen size, can't imagine legit uses outside of that…

**Comment by alloy** [comment]
I haven’t used it yet, so pretty naive on the topic of `injectGlobal`, but are there no valid reasons to use `injectGlobal` for “global rules _within a page_”?

i.e. I wouldn’t want to share global styling for one page with another, which is the case when you don’t have a SPA.

**Comment by wmertens** [comment]
For that, simply make a styled component for the page with all the rules
you want in the page. You can use any selectors you like.

**Comment by acjay** [comment]
Given that most SSR React setups I've seen involve a document frame (either as a template literal or separately rendered React component) with a stub element on which to mount the application, I'm a little confused on why one wouldn't just render a vanilla `<style>` tag into their document frame for global styles. Why should the dynamic component tree need to modify those styles? Maybe styled-components "reaching out" of their own component tree isn't the best approach for that.

**Comment by satya164** [comment]
At least in the projects I worked on, I have a single root component for the document. Any pages rendered will use the document component, but since they are different pages, they have different styles.

**Comment by geelen** [comment]
@acjay you're spot on, actually. The only issue is that React makes it hard to define a `<style>` tag inline since you need the syntax to pass through unescaped. But you could definitely do that if you preferred...

---

I'm a bit confused by your comment @satya164. If you have components right down to the document root, why do you need global styles? And, this is about SSR, so there are no browser APIs available.

I think, fundamentally, global styles break encapsulation. Every other React component only deals with itself and its children, but global styles are by definition outside your component hierarchy. For example:

```js
import reset from 'css-wipe/js'

injectGlobal`
  ${ reset }
  html { font-size: 16px; }
  @font-face {
    font-family: Gibson;
    src: url(...);
  }
`

const Root = styled.div`
  font-family: Gibson, Avenir Next, sans-serif;
  line-height: 1.15;
  font-size: 1rem;
`
```

Fonts are a good example. Declaring the font-face is a global concern, but using the font is component-specific. I'm going to push ahead with the singleton + cloning approach since it handles the specific problems with `injectGlobal` _and_ the ones with component ordering. But yep, some good alternatives in these last few replies as to why you might not want to use `injectGlobal` at all.

**Comment by satya164** [comment]
> And, this is about SSR, so there are no browser APIs available

Sure, you can just use a string if it's truly global, no APIs needed, but that's not what I want

> If you have components right down to the document root, why do you need global styles?

Because the styles can be global styles for a single page, not global styles for the all pages. Different pages reuse the same document component. I can't just set global styles on the document because that will insert the styles to all the pages, and not specific pages

**Comment by satya164** [comment]
> Every other React component only deals with itself and its children

Portals don't

**Comment by geelen** [comment]
Yeah I think the best way around that then is to really just render a `<style>` tag inline in your component hierarchy somewhere. It has the correct semantics (it has no effect once unmounted) and will be handled by React's rehydration just fine.

**Comment by ismay** [comment]
Is it recommended to start using this api if you're already on v2.0.0-15, or should I expect breaking changes? Because I'd like to start implementing ssr with v2.

**Comment by mxstbr** [comment]
@ismay `npm i --save styled-components@rc`, use the API from #679 and you're good. I doubt we'll change it again!

**Comment by mxstbr** [comment]
I'll thusly also close this issue.

**Comment by philpl** [comment]
For anyone who ends up here and is searching for some documentation. There's some text explaining it briefly on the upcoming docs: https://sc-next-docs.philpl.com/docs/advanced#server-side-rendering

**Comment by ChingxWando** [comment]
I believe the documentation was shortly updated afterwards and was live on the main site:
https://www.styled-components.com/docs/advanced#server-side-rendering

p.s. Hiii phillll & been wanting for ssr for so long & it's so exciting & thanks for creating styled component

**Comment by oyeanuj** [comment]
@philpl @wmertens Wanted to check about the FoUC part mentioned in the above discussion - is that to be expected in development mode? Or is there a way around that?

**Comment by GGAlanSmithee** [comment]
@oyeanuj as long as you render your styles server side, there won't be any FoUC, in development or production.

**Comment by mxstbr** [comment]
@oyeanuj no, there shouldn't be any FOUC. Mind opening a new issue about that?

**Comment by oyeanuj** [comment]
@mxstbr @GGAlanSmithee thanks for the quick response. It turned out to be an issue in which I had set up SC, so false alarm. Sorry for the noise!

**Comment by wilfredonoyola** [comment]
Hey :) , I have integrated this with (https://github.com/reactjs/react-rails). Actually is not working.

Can you provider some solution?

Thanks,

**Comment by philpl** [comment]
@wilfredonoyola can you please start a thread on Spectrum or open a new issue, as this one is unrelated, stale, and closed 🙂
